gobblin-api/src/main/java/gobblin/commit/CommitSequence.java

```diff
@@ -0,0 +1,146 @@
+
+
+/**
+ * A sequence of {@link CommitStep}s that should be executed atomically.
+ *
+ * <p>
+ *    Typical pattern for building a {@link CommitSequence} with two {@link CommitStep}s:
+ *
+ *    <pre>
+ *      {@code
+ *        CommitSequence sequence = CommitSequence.newBuilder()
+ *                                  .withJobName(jobName)
+ *                                  .withDatasetUrn(datasetUrn)
+ *
+ *                                  .beginStep(FsRenameCommitStep.Builder.class)
+ *                                  .withProps(props)
+ *                                  .from(srcPath)
+ *                                  .to(dstPath)
+ *                                  .endStep()
+ *
+ *                                  .beginStep(DatasetStateCommitStep.Builder.class)
+ *                                  .withProps(props)
+ *                                  .withDatasetUrn(datasetUrn)
+ *                                  .withDatasetState(datasetState)
+ *                                  .endStep()
+ *
+ *                                  .build();
+ *      }
+ *    </pre>
+ * </p>
+ *
+ * @author ziliu
+ */
+@Slf4j
+public class CommitSequence {
+
+  @Getter
+  private final String jobName;
+
+  @Getter
+  private final String datasetUrn;
+
+  private final List<CommitStep> steps;
+
+  private CommitSequence(Builder builder) {
+    this.jobName = builder.jobName;
+    this.datasetUrn = builder.datasetUrn;
+    this.steps = ImmutableList.copyOf(builder.steps);
+  }
+
+  public static class Builder {
+
+    private String jobName;
+    private String datasetUrn;
+    private final List<CommitStep> steps = Lists.newArrayList();
+
+    /**
+     * Set the job name for the commit sequence.
+     */
+    public Builder withJobName(String jobName) {
+      this.jobName = jobName;
+      return this;
+    }
+
+    /**
+     * Set the dataset URN for the commit sequence.
+     */
+    public Builder withDatasetUrn(String datasetUrn) {
+      this.datasetUrn = datasetUrn;
+      return this;
+    }
+
+    /**
+     * Build a {@link CommitStep}.
+     *
+     * @param builderClass The builder class for the {@link CommitStep}, which should extend
+     * {@link CommitStepBase.Builder}.
+     * @return An instance of the builder class for the {@link CommitStep}.
+     */
+    public <T extends CommitStepBase.Builder<?>> T beginStep(Class<T> builderClass) {
+      try {
+        //return (T) ConstructorUtils.invokeConstructor(builderClass, this);
+        return (T) builderClass.getDeclaredConstructor(this.getClass()).newInstance(this);
+      } catch (ReflectiveOperationException e) {
+        throw new RuntimeException("Failed to instantiate " + builderClass, e);
+      }
+    }
+
+    /**
+     * Add a {@link CommitStep} to the commit sequence.
+     */
+    public Builder addStep(CommitStep step) {
+      this.steps.add(step);
+      return this;
+    }
+
+    public CommitSequence build() {
+      return new CommitSequence(this);
```

Comment:
The build() method should do validation of the builder variables, e.g. name and datasetUrn should not be null. Is it acceptable to have no commit steps?